### PR TITLE
Make GC addrof and sizeof nogc

### PR DESCRIPTION
**Pull request by edi33416**
Add `@nogc` to the actual `GC` functions, as I have forgotten in #1954 

@andralex

### CODE REVIEW

File: src/core/memory.d [hunk 1]
```diff
@@ -595,14 +595,14 @@ struct GC
      * Returns:
      *  The base address of the memory block referenced by p or null on error.
      */
-    static inout(void)* addrOf( inout(void)* p ) nothrow /* FIXME pure */
+    static inout(void)* addrOf( inout(void)* p ) pure nothrow @nogc
```
Comment (edi33416) [inline src/core/memory.d]: Any reason why pure should still be omitted?

File: src/core/memory.d [hunk 2]
```diff
@@ -621,14 +621,14 @@ struct GC
      * Returns:
      *  The size in bytes of the memory block referenced by p or zero on error.
      */
-    static size_t sizeOf( in void* p ) nothrow
+    static size_t sizeOf( in void* p ) pure nothrow @nogc
```
Comment (edi33416) [inline src/core/memory.d]: Same here

**Comment by edi33416** [comment]
@andralex should I revert #1954 and put everything in this one?

**Comment by rainers** [comment]
You should also add these new requirements to gc.interface as well as the gc implementations. The existing ones are also missing from gc.proxy.

While adding nogc to addrOf and sizeOf is very likely to be in line with the spec ("does not allocate memory on the GC heap"), that is not so clear for addRange or addRoot as this very much depends on the GC implementation.

My impression is that everyone has a different idea what nogc means. Some think that it forbids any call into the GC so you can link without it (I believe Walter is in that group). Others consider not running a collection to be the crucial point, but even malloc might not do this for a concurrent GC. 

Is there a consensus on this?

File: src/core/memory.d (hunk 1 above)
Comment (rainers) [reply src/core/memory.d]: With the current rules for pure, the second call could be elided by the compiler:
```
    void* p = GC.malloc(10000);
    void* q = GC.addrOf(p);
    GC.free(p);
    void* r = GC.addrOf(p);
    assert(r is null);
```

Comment (edi33416) [reply src/core/memory.d]: So I should keep the `/* FIXME pure */`.

**Comment by edi33416** [comment]
@rainers thank you for pointing out the missing spots.

Regarding the `nogc` meaning, I don't know whether there is a consensus on this, but imho a function that does not produce garbage should be marked as `nogc`.

**Comment by ZombineDev** [comment]
Since @andralex merged the other PR, I think he's on the same page w.r.t. the meaning of `@nogc`.

**Comment by andralex** [comment]
>Is there a consensus on this?

nogc = "does not allocate memory that will need the garbage collector to free it"

@edi33416 we're good

**Comment by rainers** [comment]
> The meaning of @nogc is tied to the run-time promise of no garbage collections

GC.disable could make code nogc, then?

> nogc = "does not allocate memory that will need the garbage collector to free it"

Thanks for clarifying, that's another interpretation ;-) A matching pair of GC.malloc and GC.free will satisfy it, too, though.

I guess a closed PR is not a good place to discuss this. I might start a forum thread instead.

**Comment by andralex** [comment]
>A matching pair of GC.malloc and GC.free will satisfy it, too, though.

Affirmative, that makes a function `@nogc`. The compiler doesn't check that, so you'd need to cast the function type.

**Pull request merged**